nova/api/openstack/compute/sap_admin_api.py

```diff
@@ -69,6 +72,46 @@ def in_cluster_vmotion(self, req, body):
             common.raise_http_conflict_for_instance_invalid_state(
                 state_error, 'in_cluster_vmotion', server_id)
 
+    @_register_endpoint('GET')
+    def usage(self, req):
+        """Call
+        """
+        context = req.environ['nova.context']
+        context.can(sap_policies.POLICY_ROOT % 'usage:list')
+        return self._usage(context)
+
+    def show_usage(self, req, id):
+        """Call
+        """
+        context = req.environ['nova.context']
+        context.can(sap_policies.POLICY_ROOT % 'usage:show')
+
+        return self._usage(context, project_id=id)
+
+    def _usage(self, context, project_id=None):
+        results = nova_context.scatter_gather_skip_cell0(
+            context, objects.InstanceList.get_counts_by_project_and_az,
+            project_id=project_id)
+
+        aggregated = {}
+        for result in results.values():
+            if nova_context.is_cell_failure_sentinel(result):
+                continue
+            for (project_id, az), cell_usage in result.items():
+                az_usage = aggregated.get(project_id, {})
+                usage = az_usage.get(az, {})
+
+                for key, value in cell_usage.items():
+                    if key in usage:
+                        usage[key] += value
+                    else:
+                        usage[key] = value
+
+                az_usage[az] = usage
+                aggregated[project_id] = az_usage
```

Comment:
```python
az_usage = aggregated.setdefault(project_id, {})
usage = az_usage.setdefault(az, {})
```
Using `setdefault()` like above, we wouldn't need the last 2 lines.

Any reason for not using a `defaultdict(lambda: defaultdict(lambda: defaultdict(int)))`? Too ugly or does it not render to JSON properly? I think we could write it like this then:
```python
aggregated = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
…
    for (project_id, az), cell_usage in result.items():
        for key, value in cell_usage.items():
            aggregated[project_id][az][key] += value
```

Reply:
I forgot why I did use it in the first place, I think it was the debug output was ugly.
Switched to `defaultdict`  (with one level less, as I don't do the multi-project thing anymore)